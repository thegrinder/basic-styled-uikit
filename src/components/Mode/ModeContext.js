import { createContext } from 'react';

const ModeContext = createContext(undefined);

export default ModeContext;
