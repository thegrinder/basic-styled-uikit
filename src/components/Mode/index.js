export { default as ModeProvider } from './ModeProvider';
export { default as useMode } from './useMode';
