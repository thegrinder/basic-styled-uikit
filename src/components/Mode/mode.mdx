---
name: Mode
menu: Components
route: /components/mode
---

## **ModeProvider**



---



## **useMode**
