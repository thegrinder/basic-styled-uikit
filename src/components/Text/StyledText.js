import styled, { css } from 'styled-components';
import { string, bool } from 'prop-types';

import { getBaseLineHeight, calcLineHeight } from '../../theme/typography';
import { getTextSizingStyle, getTextColor, getTextCommonStyle } from './textSelectors';

const propTypes = {
  color: string.isRequired,
  sizing: string.isRequired,
  marginBottom: bool.isRequired,
};

const StyledText = styled.span`
  display: block;
  ${({
    theme,
    color,
    sizing,
    marginBottom,
  }) => {
    const baseLineHeight = getBaseLineHeight(theme);
    const sizingStyle = getTextSizingStyle(theme, sizing);
    return css`
      ${getTextCommonStyle(theme)}
      color: ${getTextColor(theme, color)};
      line-height: ${calcLineHeight(sizingStyle.fontSize, baseLineHeight)};
      margin-bottom: ${marginBottom ? `${getBaseLineHeight(theme)}rem` : '0'};
      ${sizingStyle}
    `;
  }}
`;

StyledText.propTypes = propTypes;

export default StyledText;
