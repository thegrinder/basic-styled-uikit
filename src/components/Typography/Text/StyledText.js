import styled, { css } from 'styled-components';
import { string, bool } from 'prop-types';

import { getBaseLineHeight, calcLineHeight } from '../typographySelectors';
import {
  getTextSizingStyle,
  getTextColor,
  getTextCommonStyle,
  getTextOpacity,
  getTextColorOnBg,
} from './textSelectors';

const propTypes = {
  color: string.isRequired,
  sizing: string.isRequired,
  marginBottom: bool.isRequired,
  emphasis: string.isRequired,
  ignoreBackground: bool.isRequired,
  bgColor: string,
};

const StyledText = styled.span`
  display: block;
  ${({
    theme,
    color,
    sizing,
    marginBottom,
    emphasis,
    ignoreBackground,
    bgColor,
  }) => {
    const sizingStyle = getTextSizingStyle(theme, sizing);
    return css`
      ${getTextCommonStyle(theme)}
      color: ${
        ignoreBackground
          ? getTextColor(theme, color)
          : getTextColorOnBg(theme, bgColor, color)
      };
      line-height: ${calcLineHeight(sizingStyle?.fontSize)};
      margin-bottom: ${marginBottom ? `${getBaseLineHeight(theme)}rem` : '0'};
      opacity: ${getTextOpacity(theme, emphasis)};
      ${sizingStyle}
    `;
  }}
`;

StyledText.propTypes = propTypes;

export default StyledText;
