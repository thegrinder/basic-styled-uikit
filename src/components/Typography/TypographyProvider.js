import { createProvider } from '../../helpers';

const TypographyProvider = createProvider('typography');

export default TypographyProvider;
