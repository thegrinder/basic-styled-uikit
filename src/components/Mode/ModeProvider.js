import React, { useCallback, useMemo, useState } from 'react';
import { oneOf } from 'prop-types';
import ModeContext from './ModeContext';

const propTypes = {
  initialMode: oneOf(['light', 'dark']),
};

const ModeProvider = ({ initialMode = 'light', ...rest }) => {
  const [mode, setMode] = useState(initialMode);
  const toggleMode = useCallback(
    () => setMode(prevMode => (prevMode === 'light' ? 'dark' : 'light')),
    []
  );
  const value = useMemo(
    () => ({
      mode,
      toggleMode,
    }),
    [mode, toggleMode]
  );
  return <ModeContext.Provider value={value} {...rest} />;
};

ModeProvider.propTypes = propTypes;

export default ModeProvider;
