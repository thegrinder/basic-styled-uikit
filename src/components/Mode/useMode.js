import { useContext } from 'react';
import ModeContext from './ModeContext';

const useMode = () => {
  const mode = useContext(ModeContext);
  if (!mode) {
    throw new Error('useMode can be used only within <ModeProvider /> context');
  }
  return mode;
};

export default useMode;
