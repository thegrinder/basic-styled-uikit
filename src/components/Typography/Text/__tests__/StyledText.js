import React from 'react';
import { render } from '@testing-library/react';
import { ModeProvider } from '../../../Mode';
import TypographyProvider from '../../TypographyProvider';
import StyledText from '../StyledText';

const children = 'children';

const renderComponent = (props = {}) =>
  render(
    <ModeProvider>
      <TypographyProvider>
        <StyledText
          color="primary"
          sizing="s"
          marginBottom={false}
          emphasis="high"
          ignoreBackground={false}
          {...props}
        >
          {children}
        </StyledText>
      </TypographyProvider>
    </ModeProvider>
  );

describe('<StyledText />', () => {
  it('should render correctly with children', () => {
    const {
      container: { firstChild },
    } = renderComponent();
    expect(firstChild).toBeDefined();
    expect(firstChild).toHaveTextContent(children);
    expect(firstChild).toMatchSnapshot();
  });

  it('should render <span> tag', () => {
    const {
      container: { firstChild },
    } = renderComponent();
    expect(firstChild.tagName).toEqual('SPAN');
  });
});
