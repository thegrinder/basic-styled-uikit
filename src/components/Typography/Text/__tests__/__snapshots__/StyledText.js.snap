// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StyledText /> should render correctly with children 1`] = `
<span
  class="StyledText-c4xq8q-0 isVWpO"
  color="primary"
>
  children
</span>
`;
